Integrate upstream changes

joe4dev opened the pull request:
## Motivation

The official AWS lambda-runtime-init received some upstream updates (typically merged in big blobs from their internal repository). We should update our customized `localstack` version to support the latest features (e.g., response streaming) and updates (e.g., the new Golang version). Beyond smaller doc updates, the main updates were:

* Jun 8, 2023 [chore(deps): Update to Golang 1.19 (#95)](https://github.com/aws/aws-lambda-runtime-interface-emulator/commit/bf7e2486034742b84a1a25d28478e147b5e65f06) ⇒ Easy deps upgrade to Go 1.19
* Apr 27, 2023 [feat: Pull upstream changes 2023/04 (#87)](https://github.com/aws/aws-lambda-runtime-interface-emulator/commit/c1cf0c518004962baa79bd82d5c4c2776e15073f) ⇒ Major changes to the init and internal API (6,845 additions and 2,726 deletions)

## Background

Internal docs: https://www.notion.so/localstack/Update-RIE-with-upstream-changes-for-response-streaming-18a303b503c54f958cfbee5841ef5402?pvs=4

## Changes

Integrates the upstream changes from the AWS `lambda-runtime-init` repo and adjusts the API accordingly (see documented commit log for details).

👉 Custom LocalStack changes `git diff 8b003ee53ffe758979cf64f43f6ae58363efcc2d..integrate-upstream-changes`

## Testing

Test against the LocalStack test suite.

🧪 The localstack branch `integrate-lambda-init-upstream-changes` contains some improved tests for the `request_id` propagation issue. For example: `tests.aws.services.lambda_.test_lambda.TestRequestIdHandling.test_request_id_invoke`

I tested other categories including:
* `tests.aws.services.lambda_.test_lambda.TestLambdaBaseFeatures`
* `tests.aws.services.lambda_.test_lambda.TestRequestIdHandling`

## TODO

- [x] Fix failing snapshot test due to `uuid` not being replaced in the test `tests.aws.services.lambda_.test_lambda.TestLambdaFeatures.test_invocation_with_qualifier`
- [x] Resolve TODOs (mostly coding out some logs, error handling etc)
- [x] Rebase after merging the permission PR https://github.com/localstack/lambda-runtime-init/pull/22 => Merged the `localstack` branch back as a quickfix because rebasing again is terrible with so many conflicts 😢

Commit:
docs: removing trailing whitespace (#76)

Removing the trailing whitespace in the README.md to improve the formatting and readability of the file.

Commit:
docs: adding Go version/release badges (#77)

Add Shield.io badges for Go version and GitHub release.
Useful for displaying important information in a visually appealing and easily accessible way at a glance.

Updated the license badge to standard Shield.io format.
As it appears it was incorrect pointing to the badge for 'aws-sam-local'.

Commit:
docs: adding content table (#78)

Add content table to README.md, making it more convenient for users to find the information they need and enabling them to quickly jump to specific sections.

yeongrokgim committed:
docs: Display environment variable as code snippet

Add markup to environment variable for better browsing

Commit:
docs: corrected bullet point's content spacing/indentation (#81)

The content for some bullet points are not correctly indented or spaced.
This causes text and code blocks to not be displayed properly, causing confusion for customers.

This change indents the content of the bullet points so it is now correctly displayed increasing readability.

jordan-brough committed:
docs: Add syntax highlighting, fix typos, other cleanups (#82)

* Add missing backslash in docker command

* Remove stray "`" from command

* Covert code snippet to code block for consistency

* Remove extraneous newline

* Fix invalid italics and use "_" for consistency

* Change (invalid) italics to header

Make it consistent with the preceding section that says:
"To build the emulator into your image"

Also, the stray whitespace before the final asterisk was incorrect.

* Add syntax highlighting to code blocks

* Add header to content table

dependabot[bot] committed:
chore(deps): bump golang.org/x/net from 0.1.0 to 0.7.0 (#85)

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.1.0 to 0.7.0.
- [Release notes](https://github.com/golang/net/releases)
- [Commits](https://github.com/golang/net/compare/v0.1.0...v0.7.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

valerena committed:
feat: Pull upstream changes 2023/04 (#87)

valerena committed:
chore(deps): Update to Golang 1.19 (#95)

joe4dev committed:
Merge remote-tracking branch 'upstream/develop' into integrate-upstream-changes

Integrate upstream changes. Notable conflicts:

* Makefile: simply upgrade Go version to 1.19 from upstream and use LocalStack changes
* go.mod: merge dependency updates from upstream and keep LocalStack changes
* go.sum: re-generate
* lambda/rapi/server.go: new interop import from upstream (?), otherwise not really conflicts just API updates and by-1-line-neighboring updates

joe4dev committed:
Remove unknown field CorrelationID for Invoke

Removed
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-f3273d2200ae8dda95046a3aa7643050f008af08f96b1359d8386682dd3bb5f9L69

joe4dev committed:
Remove unknown field CorrelationID for Init

Removed
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-2c1a36d379bd5ed6e893749912ac8473cc46dddf5765024d46b44da2eb56348dL192

joe4dev committed:
Add new sandbox init parameters

Added
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-2c1a36d379bd5ed6e893749912ac8473cc46dddf5765024d46b44da2eb56348dR212

joe4dev committed:
Add bootstrap to InitHandler parameters

Added
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-2c1a36d379bd5ed6e893749912ac8473cc46dddf5765024d46b44da2eb56348dR183

joe4dev committed:
Add new runtimeDomainRoot to NewBootstrapSingleCmd

Added
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-4f3b0e186aa0135ea4a0df9ea408382232a52ada5d8776b78f3391eb8e910890R152

joe4dev committed:
Fix new interop server interface (WIP)

Interface changes inspired by adaptations in MockInteropServer
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-ca7d03af9ea519754c0f107f0f63d2bbfbc18381703c4e226129cff807e9b719L29

joe4dev committed:
Remove SetTailLogOutput

Removed
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-af70cbe1459e50292214935e938de51f504915babf7cb657e7418ae7fb4e0109L153

joe4dev committed:
Use new way of getting the default interop server

Changed
https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/87/files#diff-4f3b0e186aa0135ea4a0df9ea408382232a52ada5d8776b78f3391eb8e910890R83

joe4dev committed:
Bump Golang Docker image for delve build

joe4dev committed:
Clarify ARM debugging limitations

joe4dev committed:
Fix sanbox creation nil pointer with new interop initialization

joe4dev committed:
Add status ready to LocalStack after await initialized (WIP)

joe4dev committed:
Add custom LocalStack LogsEgressAPI using logCollector

joe4dev committed:
Improve logging for runtime init initialization

joe4dev committed:
Merge branch 'localstack' into integrate-upstream-changes

joe4dev committed:
Fix missing request metadata in reserve call

joe4dev committed:
Add changelog entry for awsutil

dfangl reviewed (approved):
It's rather hard to get through everything, but our changes look good. Let's see if the LS tests pass :)

joe4dev committed:
Fix init error payload reporting

joe4dev commented:
Fixed the error reporting to use the new dedicated callback for `SendInitErrorResponse`

Other issues discovered during testing:

* It appears the test `TestLambdaRuntimesCommon.test_introspection_invoke` fails for for the runtimes `python3_7, java8, java8_al2, java11, java17, go1_x, provided, provided_al2` due the two missing environment variables `PWD=/var/task` and `SHLVL=0`. These are part of `CustomerEnvironmentVariables` but just not available anymore for some runtimes 🙃  I also updated the snapshots but nothing has changed at real AWS.
* `go1.x` runtime tests fail for me locally but that might be an issue with the ARM build. Both the previous init and these changes yield an error `2023-10-17 21:20:08 time="2023-10-17T19:20:08Z" level=warning msg="Process 19(bootstrap) exited: Runtime exited with error: signal: trace/breakpoint trap" func="go.amzn.com/lambda/core.(*Watchdog).GoWait.func1" file="/home/runner/work/lambda-runtime-init/lambda-runtime-init/lambda/core/watchdog.go:67"`
* Several more tests fail on macOS, most of them due to port in use when using multiple Lambda functions. Including: `test_lambda_different_iam_keys_environment`, `test_runtime_introspection_arm` (due to different debug env), `test_lambda_concurrency_block`, `test_provisioned_concurrency`, `test_reserved_concurrency_async_queue`, `test_lambda_versions_with_code_changes`, `test_lambda_alias_moving`, `test_alias_routingconfig`

joe4dev commented:
> It's rather hard to get through everything, but our changes look good. Let's see if the LS tests pass :)

Yeah, I think we just need to release an imperfect "-dev" version to get a good overview from a LocalStack CI run. (we could consider short-lived feature-branch artifacts in the future if velocity becomes a challenge)
